xpartition.py

```diff
@@ -271,3 +301,77 @@ def mappable_write(
         return functools.partial(
             _write_partition_dataset, self._obj, store, ranks, dims
         )
+
+
+def zeros_like_dataarray(arr, chunks):
+    dask_chunks = {
+        arr.get_axis_num(dim): size for dim, size in chunks.items() if dim in arr.dims
+    }
+    return xr.apply_ufunc(
+        dask.array.zeros_like, arr, kwargs=dict(chunks=dask_chunks), dask="allowed"
+    )
+
+
+def zeros_like(ds: xr.Dataset, chunks):
+    return ds.apply(zeros_like_dataarray, chunks=chunks, keep_attrs=True)
+
+
+class _ValidWorkPlan:
+    """A mapping between input and output partitionings that will
+    avoid race conditions in parallel jobs
+    """
+
+    def __init__(self, partitioner, ranks: int, dims: Sequence[Hashable]):
+
+        self._partitioner = partitioner
+        self._ranks = ranks
+        self.dims = dims
+
+    @property
+    def output_chunks(self):
+        return {dim: self._partitioner._chunks[dim] for dim in self.dims}
+
+    @property
+    def input_partition(self):
+        return self._partitioner.partition(self._ranks, self.dims)
+
+
+@dataclasses.dataclass
+class PartitionMapper:
+    """Evaluate a function on each region of a partition and store the output
+    to a zarr store
+    """
+
+    plan: _ValidWorkPlan
+    func: Callable[[xr.Dataset], xr.Dataset]
+    data: xr.Dataset
+    path: str
+
+    @property
+    def dims(self):
+        return self.plan.dims
+
+    def _initialize_store(self):
+        region = self.plan.input_partition[0]
+        iData = self.data.isel(region)
+        iOut = self.func(iData)
+        full_indexers = {dim: self.data[dim] for dim in self.dims}
+
+        dims_without_coords = (set(iOut.dims) - set(iOut.indexes)) & set(self.dims)
+        for dim in dims_without_coords:
+            iOut = iOut.assign_coords({dim: iOut[dim]})
+
+        schema = zeros_like(iOut.reindex(full_indexers), chunks=self.plan.output_chunks)
```

Comment:
Is there a reason `xarray.zeros_like` does not work here?

Reply:
It cannot rechunk arrays as far as I know.

Reply:
Would this do something undesired?
```suggestion
        schema = xr.zeros_like(iOut.reindex(full_indexers).chunk(self.plan.output_chunks))
```

Reply:
I think you suggestion is good enough, although the dask graph of the schema will have an extra layer of rechunking operations.

Reply:
I implemented your suggestion.

Reply:
Thanks -- I think `iOut.reindex(full_indexers).chunk(self.plan.output_chunks)` will have an extra layer of rechunking operations, but `schema` should just be constructed from the final chunk structure if I'm not mistaken.